accepted/single-file/design.md

```diff
@@ -0,0 +1,281 @@
+
+For example, if the TPA without single-exe publish is:
+`~/app/il.dll`; `~/app/r2r.dll`; `/usr/local/nuget/dep.dll`
+
+Where
+
+* `il.dll` is a pure managed assembly that is loaded directly from the bundle
+* `r2r.dll` is a ready-to-run assembly that is bundled and extract out to disk
+* `dep.dll` is an assembly that is not bundled with the app,
+
+the TPA with single-exe publish will be (assuming `?` is the bundle marker):
+`?/il.dll`; `/tmp/.net/e53xf3/r2r.dll`; `/usr/local/nuget/dep.dll`
+
+## Testing
+
+* Unit Tests to achieve good coverage on apps using managed, ready-to-run, native code 
+* Tests to verify that framework-dependent and self-contained apps can be published as single file.
+* Tests for ensure that every app model template supported by .NET Core (console, wpf, winforms, web, etc.) can be published as a single file.
+* A subset of CoreCLR Tests
+* End-to-End testing on real world apps such as Roslyn, MusicStore
+* *Measurements*: Publish size and run-time (first run, subsequent runs) for HelloWorld, Roslyn and MusicStore.
+
+## Further Work
+
+#### Bundler Optimizations
+
+Since all the files of an app published as a single-file live together, we can perform the following optimizations
+
+- R2R compile the app and all of its dependent assemblies in a single version-bubble
+- Investigate whether collectively signing the files in an assembly saves space for certificates.
+
+#### APIs to access Bundled files
+
+The binaries that are published in the project are expected to be handled transparently by the host. However, explicit access to the embedded files is useful in situations such as:
+
+- Reading additional files packaged into the app (ex: data files).
+- Open an assembly for reflection/inspection
+- Load plugins built as single-file class-libs using existing Loader APIs.
+
+In facilitate this usage, we can add an API similar to [GetManifestResourceStream](https://docs.microsoft.com/en-us/dotnet/api/system.reflection.assembly.getmanifestresourcestream?view=netframework-4.7.2#System_Reflection_Assembly_GetManifestResourceStream_System_String_) to obtain a stream corresponding to an embedded file. 
+
+```C#
+// Open a file embedded in the bundle built for the specified assembly 
+namespace System.Runtime.Loader
+{
+    public partial class Bundle
+    {
+        public static System.IO.Stream GetFileStream(Assembly assembly, string name);
+    }
+}
+```
+
+We can also provide an abstraction that abstracts away the physical location of a file (bundle or disk). For example, add a variant of  `GetFileStream` API that looks for a file in the bundle, and if not found, falls back to disk-lookup.
+
+#### Single-file Plugins
+
+The above design should be extended to seamlessly support single-file publish for plugins.
+
+- The bundler tool will mostly work as-is, regardless of whether we publish an application or class-lib. The binary blob with dependencies can be appended to both native and managed binaries.
```

Comment:
What happens when plugins require a common dependency that would already be there with the application? Would they have to bundle it again?
For example, let’s assume application `Foo.exe` uses a library `Foo.Common.dll`. Now, plugins can be built that use the `Foo.Common.dll`. The application then has some mechanism to pick these up.

With the classic approach, or even .NET Framework, the application might have exactly those two outputs: `Foo.exe` and `Foo.Common.dll`. And when you create the plugin, you might end up with `plugin.dll` and `Foo.Common.dll`. Now since the common library would already be bundled with the application, distributions of the plugin would only have to contain the `plugin.dll`. You could paste that into the application folder, enable the plugin, and it would work.

With .NET Core and single-file publishing, this would now mean that both the application and the plugin would bundle `Foo.Common.dll` and all the runtime dlls into their final output. So you have those twice for this setup. Of course, you could opt in to make the library _not_ bundle into a single file, but the plugin might also have some other dependencies, so it likely would _want_ to bundle into a single file. So in the end, this would require a lot of configuration for the plugin author to produce an efficient output.

Do you have any thoughts on this situation?

Reply:
This problem of what dependencies to include with plugins is not specific to single .exe files. You are right that one needs to be aware what comes with the app vs. what needs be included with the plugin. 

Take a look at plugin sample https://github.com/dotnet/samples/tree/master/core/extensions/AppWithPlugin that shows how it can be done.

Reply:
I agree, the goal of this work is to just package the contents of the `publish` directory to a single file.  It shouldn't change the contents of the publish directory. The way to resolve missing dependencies for a plugin depends on it's configuration and isolation settings. 

For example, if we didn't package `Foo.Common.dll` with `plugin.dll`, and `plugin.dll` is loaded in it's own `AssemblyLoadContext`, then it could:
* Choose to not load `Foo.Common.dll` in its own ALC, in which case, the load-logic falls back to the default ALC (of `Foo.exe`) and loads `Foo.Common.dll` in that ALC.
* Choose to explicitly load `Foo.Common.dll` from the `Foo.exe` bundle, using `GetFileStream` API proposed [here](https://github.com/dotnet/designs/pull/52/files#diff-28dba553f5f12520afc2d90d1587ef56R185).

Reply:
> This problem of what dependencies to include with plugins is not specific to single .exe deployment.

The difference for non single file deployments is that you can simply diff what files would be added by your plugin and only distribute those.

> […] the load-logic falls back to the default ALC (of `Foo.exe`) and loads `Foo.Common.dll` in that ALC.

So it would work if you deploy the app as a single file and choose not to do that with the plugin (distributing only the `plugin.dll`), and the `plugin.dll` could then still (automatically) access the common library and (more importantly) the framework dependencies from the `Foo.exe`?

The only problem I would see if the plugin has additional dependencies and wants to do a single-file deployment of only those dependencies that are specific to the plugin (and aren’t already included in the app). But I guess this would work with customization of the bundling behavior, so all the framework assemblies are not bundled.

On that note, I could actually imagine just bundling the framework assemblies for every single self-contained application, so that there’s only the .exe, the concrete dependencies, and then maybe a `framework.dll`. – Would that work somehow or would that already be problematic because the executing assembly needs to be a host file that is able to unpack the files?

Reply:
@poke : A few points to clarify:
* The plugin and app should be developed with a proper expectation of each other's dependencies (native binaries and assemblies). 
* If diffing published files is desirable, we still have the option to publish without the single-file option. The dependencies should not be different whether we publish single-file or not.
* Framework is typically never packaged with plugins.
* The runtime will provide call-backs extract file embedded within plugins. We may be able to refactor the host code such that the callbacks may reuse that code for handling single-file plugins.